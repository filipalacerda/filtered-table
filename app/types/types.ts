export type Product = {
  id: number;
  property_values: {
    property_id: number;
    value: string | number;
  }[];
};

export type Operator = {
  text: string;
  id: string;
};

export type Products = Product[];

export type Property = {
  id: number;
  name: string;
  type: "string" | "number" | "enumerated";
  values?: string[];
};

export type Filters = {
  operator: Operator;
  property: Property;
  value: string;
};
