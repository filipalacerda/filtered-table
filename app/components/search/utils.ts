import { Operator, PropertyType } from "@/app/types/types";

const categoryOperatorsMap = {
  string: ["equals", "any", "none", "in", "contains"],
  number: ["equals", "greater_than", "less_than", "any", "none", "in"],
  enumerated: ["equals", "any", "none", "in"],
};

/**
 * Not all operators are available for each property type.
 *
 * This function maps the available operators for each property type
 *
 * @param operators
 * @param currentCategoryProperty
 * @returns array
 */
const filterOperators = (
  operators: Operator[],
  currentCategoryProperty?: PropertyType
) => {
  let result: [] | Operator[];
  const availableOperators =
    currentCategoryProperty && categoryOperatorsMap[currentCategoryProperty];

  if (availableOperators) {
    result = operators.reduce((acc: Operator[], value: Operator) => {
      if (availableOperators.includes(value.id)) {
        acc.push(value);
      }
      return acc;
    }, []);
  } else {
    result = [];
  }
  return result;
};

export { filterOperators };
