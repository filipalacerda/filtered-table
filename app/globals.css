/*
* Reset Styles
*/
fieldset {
  border: none;
}

button {
  border: none;
}

.app {
  display: flex;
  padding: 6rem;
  min-height: 100vh;
  gap: 2.5rem;
  flex-direction: column;
}

@media (max-width: 768px) {
  .app {
    padding: 1rem;
  }
  .search-container {
    flex-direction: column;
  }

  .table-container {
    overflow: scroll;
  }
}

.search-container {
  display: flex;
  justify-content: space-between;
}

.table-container {
  width: 100%;
}
