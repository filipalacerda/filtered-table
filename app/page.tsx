"use client";

import { useEffect, useMemo, useState } from "react";
import datastore from "./data/data";
import Table from "./components/table";
import Search from "./components/search";

import { filterProducts } from "./filterProducts/filterProducts";

import type { Products, Operator, Property, Filters } from "./types/types";

const Home = () => {
  // datastore values
  const [products, setProducts] = useState<Products>([]);
  const [operators, setOperators] = useState<Operator[]>([]);
  const [properties, setProperties] = useState<Property[]>([]);

  // user updated values
  const [filters, setFilters] = useState<Filters>();

  useEffect(() => {
    setProducts(datastore.getProducts());
    setOperators(datastore.getOperators());
    setProperties(datastore.getProperties() as Property[]);
  }, [setProducts, setOperators, setProperties]);

  const visibleProducts = useMemo(
    () => filterProducts(products, filters),
    [products, filters]
  );

  const handleOnSearch = (values: Filters) => {
    setFilters(values);
  };

  const handleOnClear = () => {
    setFilters({});
  };

  return (
    <main className="flex min-h-screen flex-col gap-10 w-full p-24">
      <section className="flex flex-row justify-between">
        <Search
          categories={properties}
          operators={operators}
          onSubmit={handleOnSearch}
          onClear={handleOnClear}
        />
      </section>
      <section className="w-full">
        <Table headers={properties} rows={visibleProducts} />
      </section>
    </main>
  );
};

export default Home;
