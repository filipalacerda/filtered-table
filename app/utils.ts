/**
 * Given a string, it capitalizes the first letter of it
 * @param name
 * @returns string
 */
const capitalizeName = (name: string) => {
  return name.charAt(0).toUpperCase() + name.slice(1);
};

export { capitalizeName };
