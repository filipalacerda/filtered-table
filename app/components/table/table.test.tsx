import { render, screen } from "@testing-library/react";
import "@testing-library/jest-dom";
import Table from "./index";

import datastore from "../../data/data";
import { Property } from "@/app/types/types";
import { capitalizeName } from "@/app/utils";

describe(Table.name, () => {
  const headers = datastore.getProperties();
  const rows = datastore.getProducts();

  let container: HTMLElement;

  describe("when component renders", () => {
    it("should match snapshot", () => {
      container = render(
        <Table headers={headers as Property[]} rows={rows} />
      ).container;
      expect(container).toMatchSnapshot();
    });
  });

  describe("headers", () => {
    it("should render all the provided headers", () => {
      render(<Table headers={headers as Property[]} rows={rows} />);

      expect(
        screen.getByText(capitalizeName(headers[0].name))
      ).toBeInTheDocument();
      expect(
        screen.getByText(capitalizeName(headers[1].name))
      ).toBeInTheDocument();
      expect(
        screen.getByText(capitalizeName(headers[2].name))
      ).toBeInTheDocument();
      expect(
        screen.getByText(capitalizeName(headers[3].name))
      ).toBeInTheDocument();
      expect(
        screen.getByText(capitalizeName(headers[4].name))
      ).toBeInTheDocument();
    });
  });

  describe("rows", () => {
    describe("with data", () => {
      beforeEach(() => {
        container = render(
          <Table headers={headers as Property[]} rows={rows} />
        ).container;
      });

      it("should render the correct number of rows", () => {
        expect(container.querySelectorAll("tbody tr").length).toEqual(
          rows.length
        );
      });

      it("should render provided data to a row", () => {
        const firstRow = rows[0].property_values;
        expect(screen.getByText(firstRow[0].value)).toBeInTheDocument();
        // There's two `black` text being rendered in the table
        expect(screen.getAllByText(firstRow[1].value).length).toEqual(2);
        // There's two `5` text being rendered in the table
        expect(screen.getAllByText(firstRow[2].value).length).toEqual(2);
        // There's three `electronics` text being rendered in the table
        expect(screen.getAllByText(firstRow[3].value).length).toEqual(3);
        // There's two `false` text being rendered in the table
        expect(screen.getAllByText(firstRow[4].value).length).toEqual(2);
      });
    });

    describe("without data", () => {
      beforeEach(() => {
        container = render(
          <Table headers={headers as Property[]} rows={[]} />
        ).container;
      });

      it("should render only one row", () => {
        expect(container.querySelectorAll("tbody tr").length).toEqual(1);
      });

      it("should render no results message", () => {
        expect(
          screen.getByText(
            "No results found! Please try another search combination."
          )
        ).toBeInTheDocument();
      });
    });
  });
});
