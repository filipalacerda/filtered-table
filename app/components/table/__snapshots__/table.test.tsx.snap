// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table when component renders should match snapshot 1`] = `
<div>
  <table
    class="table"
  >
    <thead
      class="table-header"
    >
      <tr>
        <th
          class="header-cell"
          scope="col"
        >
          Product Name
        </th>
        <th
          class="header-cell"
          scope="col"
        >
          Color
        </th>
        <th
          class="header-cell"
          scope="col"
        >
          Weight (oz)
        </th>
        <th
          class="header-cell"
          scope="col"
        >
          Category
        </th>
        <th
          class="header-cell"
          scope="col"
        >
          Wireless
        </th>
      </tr>
    </thead>
    <tbody>
      <tr
        class="table-row"
      >
        <td
          class="cell"
        >
          Headphones
        </td>
        <td
          class="cell"
        >
          black
        </td>
        <td
          class="cell"
        >
          5
        </td>
        <td
          class="cell"
        >
          electronics
        </td>
        <td
          class="cell"
        >
          false
        </td>
      </tr>
      <tr
        class="table-row"
      >
        <td
          class="cell"
        >
          Cell Phone
        </td>
        <td
          class="cell"
        >
          black
        </td>
        <td
          class="cell"
        >
          3
        </td>
        <td
          class="cell"
        >
          electronics
        </td>
        <td
          class="cell"
        >
          true
        </td>
      </tr>
      <tr
        class="table-row"
      >
        <td
          class="cell"
        >
          Keyboard
        </td>
        <td
          class="cell"
        >
          grey
        </td>
        <td
          class="cell"
        >
          5
        </td>
        <td
          class="cell"
        >
          electronics
        </td>
        <td
          class="cell"
        >
          false
        </td>
      </tr>
      <tr
        class="table-row"
      >
        <td
          class="cell"
        >
          Cup
        </td>
        <td
          class="cell"
        >
          white
        </td>
        <td
          class="cell"
        >
          3
        </td>
        <td
          class="cell"
        >
          kitchenware
        </td>
      </tr>
      <tr
        class="table-row"
      >
        <td
          class="cell"
        >
          Key
        </td>
        <td
          class="cell"
        >
          silver
        </td>
        <td
          class="cell"
        >
          1
        </td>
        <td
          class="cell"
        >
          tools
        </td>
      </tr>
      <tr
        class="table-row"
      >
        <td
          class="cell"
        >
          Hammer
        </td>
        <td
          class="cell"
        >
          brown
        </td>
        <td
          class="cell"
        >
          19
        </td>
        <td
          class="cell"
        >
          tools
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
